### PR TITLE
Validation of UiCrossFieldChecks group on DataGrid editor commit

### PR DESCRIPTION
**Pull request by dimone-kun**
Now DataGrid will validate UiCrossFieldChecks of the edited item.  If there are validation errors DataGrid will show user error message.

### CODE REVIEW

**Comment by glebfox** [comment]
Please, include a demo project for testing purpose.

File: modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java [hunk 1]
```diff
@@ -216,6 +230,26 @@ protected void commitEditor() throws FieldGroup.CommitException {
             fireEditorPreCommitEvent();
 
             Map<Field<?>, Validator.InvalidValueException> invalidValueExceptions = commitFields();
+
+            if (isCrossFieldValidate()) {
+                @SuppressWarnings("unchecked") Entity entity = collectionDatasource.getItem(getEditedItemId());
+                javax.validation.Validator validator = beanValidation.getValidator();
+                Set<ConstraintViolation<Entity>> violations = validator.validate(entity, UiCrossFieldChecks.class);
+
+                Validator.InvalidValueException[] validationExceptions = violations.stream()
```
[the remark this answers is not in the document]
Comment (dimone-kun) [reply modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java]: Cause `com.vaadin.data.Validator.InvalidValueException#InvalidValueException(java.lang.String, com.vaadin.data.Validator.InvalidValueException...)` constructor accepts array

[the remark this answers is not in the document]
Comment (glebfox) [reply modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java]: Ok, I see

File: modules/web/src/com/haulmont/cuba/web/gui/components/WebDataGrid.java [hunk 2]
```diff
@@ -61,16 +99,29 @@
...
+import com.vaadin.ui.Grid;
+import com.vaadin.ui.HorizontalLayout;
 import org.apache.commons.collections4.CollectionUtils;
 import org.apache.commons.lang.StringUtils;
 import org.dom4j.Element;
 
 import javax.annotation.Nullable;
 import java.beans.PropertyChangeListener;
-import java.util.*;
+import java.util.ArrayList;
```
Comment (jreznot) [inline modules/web/src/com/haulmont/cuba/web/gui/components/WebDataGrid.java]: We are using optimized imports everywhere

File: modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java [hunk 3]
```diff
@@ -216,6 +230,26 @@ protected void commitEditor() throws FieldGroup.CommitException {
             fireEditorPreCommitEvent();
 
             Map<Field<?>, Validator.InvalidValueException> invalidValueExceptions = commitFields();
+
+            if (isCrossFieldValidate()) {
+                @SuppressWarnings("unchecked") Entity entity = collectionDatasource.getItem(getEditedItemId());
```
[the remark this answers is not in the document]
Comment (dimone-kun) [reply modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java]: Cause all `InvalidValueException`s are wrapped into another one exception which is accepts an array as a constructor parameter I extracted this code into the method that throws this wrapper-exception if there were exceptions of cross field validation.

Is this all right?

[the remark this answers is not in the document]
Comment (glebfox) [reply modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java]: Well, it seems that in this case, the `commitEditor` will have inconsistent behavior. I mean that there is the code that collects fields' exceptions and then check them to decide either to fire an `EditorPostCommitEvent` or to throw an exception. As far as I understand, your code behaves independently of the logic above and this makes code less readable. I propose to return an array of exceptions and then let the `commitEditor` decide what to do with all these exceptions (I mean check if exceptions are empty according to some priority and if all are empty, then fire the `EditorPostCommitEvent`).

File: modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java [hunk 4]
```diff
@@ -284,11 +264,30 @@ protected boolean isEditorFieldsValid() {
         }
     }
 
-    public boolean isCrossFieldValidate() {
+    protected void itemCrossFieldValidate() throws Validator.InvalidValueException {
```
Comment (glebfox) [inline modules/web/src/com/haulmont/cuba/web/toolkit/ui/CubaGrid.java]: Do not through an exception from this method. See the discussion above for more details.

**Comment by glebfox** [comment]
Close due to inactivity

**Pull request closed without merging**